revolv/payments/utils.py

```diff
@@ -1,6 +1,398 @@
+import datetime
+import urllib
+
+from django.db.models import Sum
+from revolv.payments.models import Payment, AdminReinvestment, AdminRepayment, AdminAdjustment
+from revolv.project.models import Project
+
 class NotEnoughFundingException(Exception):
     pass
 
 
 class ProjectNotCompleteException(Exception):
     pass
+
+class AccountingAggregator():
+    
+    """
+    An aggregator class that does much of the computation for the AdministratorAccountingView
+    view in revolv.administrator.views.
+    This class contains a bunch of helper functions and one main function (aggregate) which returns 
+    a dictionary of relevant accounting data, filtered by instance variables representing the start date,
+    end date, and project filters.
+    """
+
+    def __init__(self, start_date=None, end_date=None, project_filter=None, revolv_earnings_rate=0, payment_service_fee_rate=0):
+        # to prevent uneccessary operations
+        if start_date:
+            self.start_date = start_date
+            self.end_date = end_date
+            self.project_filter = project_filter
+            self.revolv_earnings_rate = revolv_earnings_rate
+            self.payment_service_fee_rate = payment_service_fee_rate
+            self.allProjects = Project.objects.all()
+
+    def calculate_cash_balance_from_start(self):
+        """
+        A helper function that calculates the initial cash balance from the
+        beginning of time to self.start_date
+
+        :return:
+            Returns a number that represents initial cash balance.
+        """
+        all_donations = Payment.objects.donations(end_date=self.start_date).aggregate(Sum('amount'))['amount__sum']
+        all_reinvestments = AdminReinvestment.objects.reinvestments(end_date=self.start_date).aggregate(Sum('amount'))['amount__sum']
+        all_repayments = AdminRepayment.objects.repayments(end_date=self.start_date).aggregate(Sum('amount'))['amount__sum']
+        cash_in_adjustments = AdminAdjustment.objects.adjustments(end_date=self.start_date, cash_type='cash_in').aggregate(Sum('amount'))['amount__sum']
+        cash_out_adjustments = AdminAdjustment.objects.adjustments(end_date=self.start_date, cash_type='cash_out').aggregate(Sum('amount'))['amount__sum']
+
+        if not all_donations:
```

Comment:
These aggregations will be None if there is nothing to aggregate over, I think. So you can actually just do `all_donations = all_donations or 0`